packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/graph.rs

```diff
@@ -0,0 +1,307 @@
+use std::{collections::HashMap, fmt::Debug, hash::Hash};
+
+/// An abstract graph. The speed of lookups is determined by the `Taggable` type,
+/// which is used to map `Node -> usize` for lookups in the internal data structures.
+///
+/// Example using `HashMap` as the `Taggable` type:
+/// ```rust
+/// # use doenetml_core::dep_graph::graph::Graph;
+/// use std::collections::HashMap;
+/// // Make a graph
+/// // a -> b -> c
+/// // a -> c
+/// let mut graph = Graph::<String, HashMap<_, _>>::new();
+/// graph.add_node("a".into());
+/// graph.add_node("b".into());
+/// graph.add_node("c".into());
+/// graph.add_edge("a".into(), "b".into());
+/// graph.add_edge("a".into(), "c".into());
+/// graph.add_edge("b".into(), "c".into());
+/// assert_eq!(graph.walk_descendants("a".into()).collect::<Vec<_>>(), vec!["b", "c"]);
+/// ```
+#[derive(Debug)]
+pub struct Graph<Node: Clone + Eq + Hash + Debug, Index: Taggable<Node, usize>> {
+    /// An `Taggable` that allows for looking up the index of a node in the graph.
+    /// I.e., used for `Node -> usize` lookups.
+    index: Index,
+    /// The nodes in the graph. Internally nodes are represented as `usize`. But `Graph` stores nodes
+    /// for the users to make its internal structure opaque.
+    nodes: Vec<Node>,
+    /// Directed edges between nodes.
+    edges: Vec<Vec<usize>>,
```

Comment:
Additional explanation of the data structure would be helpful here. My first thought when seeing the nested vectors was that `edges[i][j]` would have something to do with an edge from `i` to `j` rather than the `j`th edge that has been created coming from node `i`.